Reconcile core data generation features with latest research advances

bbrowning opened the pull request:
This PR brings in some of the latest advancements prototyped by our research team into the broader codebase for everyone's use. It's a work-in-progress, but also something that others may wish to follow, comment on, and contribute to as the work gets done. There are still outstanding features not yet added to this - some new pipeline block types, an improved skills pipeline config, LLMLogProbBlock and LLMMessagesBlock are just stubs, etc.

And, this may cause the deprecation / removal of some existing functionality. That is not entirely clear yet, but will become apparent as work progresses.

See `docs/upgrading_from_v0.6.x_to_v0.7.x.md` for more details, although that too is still only stubbed out.

bbrowning committed:
Remove ImportBlock as a pipeline block

While this was technically part of our public Python API, it appears
to be entirely unused. Let's pull it out now to make syncing with the
latest research advancements easier.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning committed:
Switch to Jinja2 templates for prompt templates

This stubs in support for Jinja templates in the LLMBlock prompt
templates, opening us up to more expressive prompts and handling things
like loops that take a variable number of input elements when
rendering templates.

NOTE: This is a backwards-incompatible change in prompt templates. Any
users that had custom pipelines specified will need to update their
template variables to look like `{{variable}}` instead of `{variable}`
as a result of this change.

Co-authored-by: shivchander <shivchander.s30@gmail.com>
Co-authored-by: abhi1092 <abhi1092@gmail.com>
Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning committed:
Add BlockRegistry and PromptRegistry

The Block and Prompt registries are how we keep track of what our
supported Block types are and which Prompts map to which teacher
models.

Co-authored-by: shivchander <shivchander.s30@gmail.com>
Co-authored-by: abhi1092 <abhi1092@gmail.com>
Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning committed:
Move model prompts to jinja templates and messages

This brings in changes to move our model prompt templates to Jinja
templates and the HuggingFace messages formats, used by their chat
templates.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning committed:
Stub in LLMLogProbBlock and LLMMessagesBlock

These new blocks don't do anything yet, but stubbing them into the
codebase and will continue working on figuring out what they're
supposed to do and wiring things up with tests.

Co-authored-by: shivchander <shivchander.s30@gmail.com>
Co-authored-by: abhi1092 <abhi1092@gmail.com>
Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning committed:
Add CHANGELOG.md entries for research reconciliation

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning commented:
The fact that the e2e-small test passed even though I haven't actually converted any of our default config prompt templates to Jinja syntax yet is concerning, as that means the test is extremely loose in what it considers success. We wouldn't have actually included any of the user's question or answers in the skills or knowledge prompts we sent to the model, and instead it would have all had placeholder python String format tokens in it.

bbrowning committed:
Update all knowledge configs to use jinja templates

In addition to updating the knowledge configs to use jinja templates,
this adds additional tests to validate that we are using jinja
templates instead of python string formats. That also required
tightening up our usage of jinja `Template` to always preferred
`StrictUndefined` behavior everywhere we use it.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning committed:
Update all skill configs to use jinja templates

This also makes the test running `Block._validate` on all our shipped
configs a bit more generic so that it can cover all skill and
knowledge yaml files without having to keep a separate list of config
files to test.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning committed:
Actually use BlockRegistry to lookup blocks

This gets rid of the hardcoded block types dict and drives everything
off the BlockRegistry. This means I also added a functional test
showing how users can create and register their own Block
implementations and use those in their pipeline config files - see
`tests/testdata/custom_block_pipeline.yaml` and
`tests/testdata/custom_block.py` for those examples.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning committed:
Use PromptRegistry for all chat templates

This removes the mapping of model families in SDG itself between
granite, mixtral, mistral, merlinite, etc. Instead, it uses the
PromptRegistry to lookup chat templates based on the model family
given. And, if no model family is given, it still falls back to doing
a best-guess based on the file path of the selected teacher model.

A simple test was added to demonstrate how to register and use custom
chat templates for generating prompts via the PromptRegistry.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning committed:
Port the new IterBlock from research code

This adds a new Block type - `IterBlock` - that calls another block N
times for a set of given input samples. Every iteration through the
loop, the samples returned from the child block's `generate` call get
added to the list of samples produced from this block.

So, if you use an `IterBlock` to call an `LLMBlock` 5 times, you'll get
5 samples generated (and 5 calls to the LLM) for every sample in the
source dataset. The output dataset will contain all 5 generated samples
resulting from each 1 input sample.

Co-authored-by: shivchander <shivchander.s30@gmail.com>
Co-authored-by: abhi1092 <abhi1092@gmail.com>
Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning commented:
Ok, I believe this has enough of the research code ported over, new tests created, and existing tests passing that it's ready for review. I know it's a big effort to review this, but it was also a big lift to get the core improvements from research working in our codebase.

BlockRegistry, PromptRegistry, IterBlock, and Jinja prompt templates should all be working. I created #413 and #414 separately to track finishing up `LLMLogProbBlock` and `LLMMessagesBlock`.

This does change some of our public API, although at this time there are no known users of the breaking changes there - basically removal of `ImportBlock` and reorganizing blocks under a `instructlab.sdg.blocks` package. The API used by the InstructLab CLI should be unchanged, and the e2e CI tests still passing without changes confirms this.

jwm4 reviewed (approved):
I don't see any problems with any of these changes, but I am not really an expert (and don't have write access, so my approval doesn't count for much).  I would note that the research code base also has a very nice README file with a lot of useful information.  I would like that merged in too with edits as needed to reflect any differences between that code and the open source code.   However, this PR is already plenty big, so I would recommend a separate PR for the README merge.

khaledsulayman reviewed (commented):
thanks for this massive lift!

left some minor comments, besides that everything else looks good!

bbrowning commented:
> I don't see any problems with any of these changes, but I am not really an expert (and don't have write access, so my approval doesn't count for much). I would note that the research code base also has a very nice README file with a lot of useful information. I would like that merged in too with edits as needed to reflect any differences between that code and the open source code. However, this PR is already plenty big, so I would recommend a separate PR for the README merge.

I agree we should pull in those changes - tracked as #428.

bbrowning commented on src/instructlab/sdg/blocks/block.py:
```diff
@@ -0,0 +1,65 @@
+
+        :param prompt_template: The Jinja2 template object.
+        :param input_dict: A dictionary of input values to check against the template.
+        :return: True if the input data is valid (i.e., no missing variables), False otherwise.
```
Yes, great catch - I just copied the comments over verbatim, but pushed a new commit to adapt these to the style you outlined that we use elsewhere in SDG.

bbrowning commented on src/instructlab/sdg/blocks/block.py:
```diff
@@ -0,0 +1,65 @@
+            # Try rendering the template with the input_dict
+            prompt_template.render(ChainMap(input_dict, Default()))
+            return True
+        except UndefinedError as e:
```
You made me realize some of this is leftover from pre-Jinja stuff - we can do away with KeyError and ChainMap entirely here, and just rely on Jinja to throw UndefinedError for us on any undefined keys. Simplifying this logic to do that in another commit.

aakankshaduggal commented on src/instructlab/sdg/blocks/iterblock.py:
```diff
@@ -0,0 +1,42 @@
+
+# This is part of the public API.
+@BlockRegistry.register("IterBlock")
+class IterBlock(Block):
```
Could we add a docstring to explain what this new block does? It would be helpful to briefly describe its purpose, key attributes like `num_iters`, and how it fits into the overall system. This will make it easier for others to understand its functionality and use. We also don't have to add it in this PR, just a recommendation.

bbrowning replied:
Done!

aakankshaduggal reviewed (approved):
Thanks @bbrowning for a great PR! LGTM 🚢

anastasds commented on src/instructlab/sdg/blocks/llmblock.py:
```diff
@@ -300,6 +290,13 @@ def __init__(
+        for config_path in config_paths:
+            assert (
+                len(config_path) == 2
+            ), "ConditionalLLMBlock config_paths each entry should be a list of config path and selector column names"
```
What's the motivation to use `assert` here instead of raising?

bbrowning replied:
We typically only use `assert` outside of tests to validate/document code assumptions that are not meant to be user-facing or influenced by user config or params. However, in this case, I think you're properly catching that these config_paths are in fact user-facing (as they specify them in their pipeline yaml) and thus these should be a raised exception instead. Will fix :smile:

bbrowning replied:
This now raises a `BlockConfigParserError` instead of asserting, and adds a test to validate that gets raised properly. Thanks for the input!

bbrowning committed:
Validate blocks by raising BlockConfigParserError instead of asserts

Asserts outside of tests should only be used for programming errors in
our own code and not to validate user-facing things.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning commented:
Rebased on top of latest main since CI is now fixed there - will let CI chew on things and if it's good and no more comments, going to squash a lot of these into a handful of fewer commits, add the co-authorship metadata for the upstream researchers that did much of the initial code, and get this merged. I'm not planning any functional changes to this PR from this point forward unless more review notes come up.
